File: Java/Loon-Neo/src/loon/opengl/LTextureBind.java
/**
 * Copyright 2008 - 2015 The Loon Game Engine Authors
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.5
 */
package loon.opengl;

import loon.LTexture;
import loon.utils.GLUtils;

public class LTextureBind extends GLBase {

	public static abstract class Source {

		public abstract String fragmentShader() ;
		
		public abstract String vertexShader() ;
		
	}

	public final GL20 gl;
	protected int curTexId = -1;
	protected int lastTexId = -1;

	public int getCurrentTextureID() {
		return this.curTexId;
	}

	public void setTexture(final LTexture texture) {
		final int id = texture.getID();
		if (!texture.isLoaded()) {
			texture.loadTexture();
		}
		if (curTexId != -1 && curTexId != id) {
			flush();
		}
		this.lastTexId = this.curTexId;
		this.curTexId = id;
	}

	@Override
	public void end() {
		super.end();
		lastTexId = -1;
		curTexId = -1;
	}

	protected LTextureBind(GL20 gl) {
		this.gl = gl;
	}

	protected void bindTexture() {
		GLUtils.bindTexture(gl, curTexId);
	}

	@Override
	public void init() {

	}

}

File: C#/Loon2MonoGame/LoonMonoGame-Lib/loon/opengl/LTextureBind.cs
using loon.utils;

namespace loon.opengl
{
	public class LTextureBind : GLBase
	{

		public readonly GL20 gl;
		protected internal int curTexId = -1;
		protected internal int lastTexId = -1;

		public virtual int CurrentTextureID
		{
			get
			{
				return this.curTexId;
			}
		}

		public virtual LTexture Texture
		{
			set
			{
				//JAVA TO C# CONVERTER WARNING: The original Java variable was marked 'final':
				//ORIGINAL LINE: final int id = value.getID();
				int id = value.ID;
				if (!value.IsLoaded())
				{
					value.LoadTexture();
				}
				if (curTexId != -1 && curTexId != id)
				{
					Flush();
				}
				this.lastTexId = this.curTexId;
				this.curTexId = id;
			}
		}

		public override void End()
		{
			base.End();
			lastTexId = -1;
			curTexId = -1;
		}

		protected internal LTextureBind(GL20 gl)
		{
			this.gl = gl;
		}

		protected internal virtual void BindTexture()
		{
			GLUtils.BindTexture(gl, curTexId);
		}

		public override void Init()
		{

		}

	}

}
